io-engine/src/grpc/v1/stats.rs

```diff
@@ -0,0 +1,200 @@
+use crate::{
+    core::lock::ResourceLockManager,
+    grpc::{
+        rpc_submit,
+        v1::{pool::PoolService, replica::ReplicaService},
+        GrpcClientContext,
+        GrpcResult,
+        ReadLock,
+        Serializer,
+    },
+    lvs::Lvs,
+};
+use futures::FutureExt;
+use io_engine_api::v1::stats::*;
+use std::fmt::Debug;
+use tonic::{Request, Response, Status};
+
+use crate::core::{BlockDeviceIoStats, CoreError, UntypedBdev};
+use ::function_name::named;
+use std::panic::AssertUnwindSafe;
+
+/// RPC service for Resource IoStats.
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct StatsService {
+    name: String,
+    client_context:
+        std::sync::Arc<tokio::sync::RwLock<Option<GrpcClientContext>>>,
+    pool_svc: PoolService,
+    replica_svc: ReplicaService,
+}
+
+#[async_trait::async_trait]
+impl<F, T> Serializer<F, T> for StatsService
+where
+    T: Send + 'static,
+    F: core::future::Future<Output = Result<T, Status>> + Send + 'static,
+{
+    async fn locked(&self, ctx: GrpcClientContext, f: F) -> Result<T, Status> {
+        // Takes read lock of Pool and Replica service.
+        let pool_context = self.pool_svc.client_context.read().await;
+        let replica_context = self.replica_svc.client_context.read().await;
+
+        if let Some(c) = pool_context.as_ref() {
+            warn!("{}: gRPC method timed out, args: {}", c.id, c.args);
+        }
+        if let Some(c) = replica_context.as_ref() {
+            warn!("{}: gRPC method timed out, args: {}", c.id, c.args);
+        }
+        // Taking write lock of Stats service. This will hold off read ops.
+        let _statsvc_lock = self.client_context.write().await;
+
+        let lock_manager = ResourceLockManager::get_instance();
+        // For nexus global lock.
+        let _global_guard =
+            match lock_manager.lock(Some(ctx.timeout)).await {
+                Some(g) => Some(g),
+                None => return Err(Status::deadline_exceeded(
+                    "Failed to acquire access to object within given timeout",
+                )),
+            };
+        let fut = AssertUnwindSafe(f).catch_unwind();
+        let r = fut.await;
+
+        match r {
+            Ok(r) => r,
+            Err(_e) => {
+                warn!("gRPC method panicked, args");
+                Err(Status::cancelled(format!("gRPC method panicked")))
+            }
+        }
+    }
+}
+
+impl StatsService {
+    async fn executor<
+        T: Send + 'static,
+        F: core::future::Future<Output = Result<T, Status>> + Send + 'static,
+    >(
+        &self,
+        _reader: tokio::sync::RwLockReadGuard<'_, Option<GrpcClientContext>>,
```

Comment:
Better to pass reader, and then lock self.client_context first and only then lock the reader, otherwise you're unnecessarily locking the pool_service

```suggestion
         reader: &impl RWLock,
```